### PR TITLE
Cico planner

### PR DESCRIPTION
**Pull request by pranavgore09**
Run build + unit + func tests on cico

### CODE REVIEW

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-3 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-4 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

File: cico_run_tests.sh [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+# Build builder image
+cp /tmp/jenkins-env .
+
+mkdir -p fabric8-ui-dist
+# Build fabric8-planner image
+docker build -t fabric8-planner-builder -f Dockerfile .
+# User root is required to run webdriver-manager update.
+# This shouldn't be a problem for CI containers
+docker run --detach=true --name=fabric8-planner -v $(pwd)/fabric8-ui-dist:/home/fabric8/fabric8-planner/fabric8-ui-dist:Z --cap-add=SYS_ADMIN -t fabric8-planner-builder
```
Comment (kwk) [inline cico_run_tests.sh]: `docker run` returns the ID of the container being run. It would be nice if you don't specify the name `fabric8-planner` but instead use the ID below in `docker exec fabric8-planner`.

File: cico_run_tests.sh [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+docker exec fabric8-planner npm run tests -- --unit
+
+## Exec functional tests
+docker exec fabric8-planner bash -c 'cd runtime; npm install'
+docker exec fabric8-planner bash -c 'cd runtime/src/tests/functionalTests; DEBUG=true HEADLESS_MODE=true ./run_ts_functional_tests.sh smokeTest'
+
+# Following steps will create a snapshot for testing
+
+# Build and integrate planner with fabric8-ui
+docker exec fabric8-planner npm pack dist/
```
Comment (kwk) [inline cico_run_tests.sh]: Instead of running each command individually you might as well mount a file to the `fabric8-planner` image and execute that. It probably is cleaner.

File: fabric8-ui-dist/package.json [hunk 3]
```diff
@@ -0,0 +1,215 @@
+{
+  "name": "fabric8-planner",
```
Comment (kwk) [inline fabric8-ui-dist/package.json]: Why is this file added here? It looks pretty much like something that exists in another place as well and maintaining multiple versions in different locations is too hard.

File: fabric8-ui-dist/package.json [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+  },
+  "bugs": {
+    "url": "https://github.com/fabric8-ui/fabric8-planner/issues"
+  },
+  "engines": {
+    "node": ">= 8.3.0",
+    "npm": ">= 5.3.0"
+  },
+  "dependencies": {
+    "@angular/animations": "^4.4.0",
```
Comment (kwk) [inline fabric8-ui-dist/package.json]: Don't you want to use sticky versions? @michaelkleinhenz what do you think?

File: fabric8-ui-dist/package.json (hunk 3 above)
Comment (pranavgore09) [reply fabric8-ui-dist/package.json]: accident!

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-5 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-6 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-7 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-8 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-10 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-11 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-12 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-13 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-14 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-15 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-16 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-22 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by debloper** [comment]
[test]

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-39 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-40 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-41 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

File: Dockerfile [hunk 5]
```diff
@@ -48,6 +48,10 @@ RUN curl -SLO "https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION-linux-
 #  && npm install -g karma-firefox-launcher
 
 COPY runtime/tests/google-chrome.repo /etc/yum.repos.d/google-chrome.repo
+
+# following downgrade is needed as GKE build does not install latest chrome.
+# Functional tests do pass on latest chrome version-65 (as of this writing)
+# Replace following 2 lines by following whenever you want to upgrade to latest
```
Comment (jarifibrahim) [inline Dockerfile]: Typo here. It should be
`Remove the following 2 lines whenever you wish to upgrade to the latest version of chrome`

File: cico_run_tests.sh [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+# Build builder image
+cp /tmp/jenkins-env .
+
+mkdir -p fabric8-ui-dist
+# Build fabric8-planner image
+docker build -t fabric8-planner-builder -f Dockerfile .
+# User root is required to run webdriver-manager update.
+# This shouldn't be a problem for CI containers
+# Chrome crashed on low size of /dev/shm. We need the --shm-size=256m flag.
```
Comment (jarifibrahim) [inline cico_run_tests.sh]: Typo here (I made it :D). It should be
`Chrome crashes on low size of /dev/shm. We need the --shm-size=256m flag`

**Comment by pranavgore09** [comment]
This PR is now working on both platforms cico and f8.jenkins.
The only change I need to make is google chrome getting downloaded from random repo [here](https://github.com/fabric8-ui/fabric8-planner/pull/2473/files#diff-3254677a7917c6c01f55212f86c57fbfR56)
I am struggling to find out official chrome releases for each version. Once that is done this PR will be good to merge.

**Comment by debloper** [comment]
> We should separate those out and provide a shell-oriented Makefile that one can execute on a local machine as well.

@kwk agreed, but can that be considered into a followup patch? We (me and pranav) are considering addition of Makefile and define its most appropriate purpose (without taking-away/stepping-onto npm or gulp's tails). But that will need a bit of followup, and better done when this and the standalone Planner patch are through (so to say, the architecture is a bit more stable).

**Comment by kwk** [comment]
@debloper I'm just here to say that npm and gulp can be as good as they want but they probably will never beat a Makefile when it comes to just executing commands and making the computer do something. So please don't put this build stuff into npm or gulp would you?

File: cico_run_tests.sh [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    | sed 's/^/export /g' \
+    > /tmp/jenkins-env
+  source /tmp/jenkins-env
+fi
+
+# We need to disable selinux for now, XXX
+/usr/sbin/setenforce 0
+
+# Get gcc-c++ to avoid npm errors
+yum -y install gcc-c++
```
Comment (kwk) [inline cico_run_tests.sh]: @pranavgore09 @debloper why is that installation needed here and why isn't that done in the Docker image that is used below?

File: cico_run_tests.sh [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+# Get all the deps in
+yum -y install docker make git
```
Comment (kwk) [inline cico_run_tests.sh]: What do you need git for on the host that runs this script @pranavgore09 @debloper ?

Comment (kwk) [inline cico_run_tests.sh]: What do you need git for on the host that runs this script @pranavgore09 @debloper ?

File: cico_run_tests.sh (hunk 7 above)
Comment (kwk) [inline cico_run_tests.sh]: @pranavgore09 @debloper why is that installation needed here and why isn't that done in the Docker image that is used below?

File: cico_run_tests.sh (hunk 8 above)
Comment (debloper) [reply cico_run_tests.sh]: Don't think `git` and `make` either is required at the moment.

File: cico_run_tests.sh [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+service docker start
+
+# Build builder image
+cp /tmp/jenkins-env .
+
+mkdir -p fabric8-ui-dist
+# Build fabric8-planner image
+docker build -t fabric8-planner-builder -f Dockerfile .
```
Comment (debloper) [inline cico_run_tests.sh]: `-f Dockerfile` is default and can be safely ignored.
`-f` is only needed to be mentioned for Dockerfiles with different names.

File: cico_run_tests.sh (hunk 7 above)
Comment (debloper) [reply cico_run_tests.sh]: More than half the content of this file should actually be part of the `Dockerfile` IMO
FWIW, the Dockerfile is even missing the `ENTRYPOINT/CMD` and that's not good.

**Comment by debloper** [comment]
- The spec-file changes shouldn't need to be part of this PR.
- Put stuff back into the Dockerfile instead of running those steps with `docker exec`
- If you need to run multiple commands as entrypoint, then make a script out of them and run that.

File: cico_run_tests.sh (hunk 8 above)
Comment (pranavgore09) [reply cico_run_tests.sh]: we are cloning fabric8-ui using `git`. we need to clone to build snapshot image.

File: cico_run_tests.sh (hunk 7 above)
Comment (pranavgore09) [reply cico_run_tests.sh]: to avoid npm warnings, see discussions at https://github.com/fabric8io-images/fabric8-ui-builder/pull/6

**Comment by pranavgore09** [comment]
thanks @kwk for spending time on review and for creating this https://github.com/fabric8-ui/fabric8-planner/issues/2506 issue for adding makefile.
I just want to keep this PR small and short, already we are having so many abstraction levels to find from. Sometimes I feel the single source of truth is better than jumping all over places. Still, I agree with your concerns and I would like to keep #2506  out of this PR.

File: cico_run_tests.sh (hunk 8 above)
Comment (pranavgore09) [reply cico_run_tests.sh]: removed as `git` is present in Dockerfile and `make` not needed.
https://github.com/fabric8-ui/fabric8-planner/pull/2473/commits/4139eb103b8ea1ae3f5eb59cf716b983624208d2

File: cico_run_tests.sh (hunk 7 above)
Comment (pranavgore09) [reply cico_run_tests.sh]: moved to Dockerifle https://github.com/fabric8-ui/fabric8-planner/pull/2473/commits/4139eb103b8ea1ae3f5eb59cf716b983624208d2

File: cico_run_tests.sh (hunk 9 above)
Comment (pranavgore09) [reply cico_run_tests.sh]: thanks done in https://github.com/fabric8-ui/fabric8-planner/pull/2473/commits/537acb7716f3dd17ce1e0ae13a669e45edb24da0

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-49 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by jarifibrahim** [comment]
@debloper 

> The spec-file changes shouldn't need to be part of this PR.

Spec file changes are absolutely necessary and should go in with this PR. Those changes stabilize the tests.

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-51 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-52 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-53 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-54 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

File: cico_run_tests.sh [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+mkdir -p fabric8-ui-dist
+# Build fabric8-planner image
+docker build -t fabric8-planner-builder .
+# User root is required to run webdriver-manager update.
+# This shouldn't be a problem for CI containers
+# Chrome crashes on low size of /dev/shm. We need the --shm-size=256m flag.
+CID=$(docker run --detach=true \
+    --shm-size=256m \
+    -v $(pwd)/fabric8-ui-dist:/home/fabric8/fabric8-planner/fabric8-ui-dist:Z \
```
Comment (kwk) [inline cico_run_tests.sh]: Please change to: `-v $(pwd)/fabric8-ui-dist:/home/fabric8/fabric8-planner/dist:Z \`

File: cico_run_tests.sh [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    export PROXIED_K8S_API_SERVER="${WS_K8S_API_SERVER}"
+    export OAUTH_ISSUER="https://${WS_K8S_API_SERVER}"
+    export PROXY_PASS_URL="https://${WS_K8S_API_SERVER}"
+    export OAUTH_AUTHORIZE_URI="https://${WS_K8S_API_SERVER}/oauth/authorize"
+    export AUTH_LOGOUT_URI="https://${WS_K8S_API_SERVER}/connect/endsession?id_token={{id_token}}"
+
+    cd fabric8-ui && npm run build:prod
+'''
+# Copy dist and Dockerfile.deploy to host (via mounted dir)
+docker exec $CID bash -c 'cd fabric8-ui; cp -r dist/ /home/fabric8/fabric8-planner/fabric8-ui-dist/'
```
Comment (kwk) [inline cico_run_tests.sh]: No need to copy because it is already mounted to the correct place (see https://github.com/fabric8-ui/fabric8-planner/pull/2473/files#r176723542).

File: cico_run_tests.sh [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+# Build builder image
+cp /tmp/jenkins-env .
+
+mkdir -p fabric8-ui-dist
+# Build fabric8-planner image
+docker build -t fabric8-planner-builder .
+# User root is required to run webdriver-manager update.
+# This shouldn't be a problem for CI containers
+# Chrome crashes on low size of /dev/shm. We need the --shm-size=256m flag.
+CID=$(docker run --detach=true \
```
Comment (kwk) [inline cico_run_tests.sh]: Maybe add `-u $(shell id -u $(USER)):$(shell id -g $(USER))`?

File: cico_run_tests.sh (hunk 10 above)
Comment (pranavgore09) [reply cico_run_tests.sh]: hi @kwk while we were discussing I forgot, this `dist` we want on the host is not from planner. It is from fabric8-ui directory that we have cloned inside the container. So, I do not see other option than to copy it ATM

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-55 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

File: cico_run_tests.sh (hunk 11 above)
Comment (pranavgore09) [reply cico_run_tests.sh]: pease see https://github.com/fabric8-ui/fabric8-planner/pull/2473#discussion_r176928132

File: cico_run_tests.sh (hunk 10 above)
Comment (kwk) [reply cico_run_tests.sh]: @pranavgore09 you could simply mount two directories into the container :)

Comment (debloper) [reply cico_run_tests.sh]: Are we talking about this: https://github.com/fabric8-ui/fabric8-planner/blob/master/scripts/bootstrap.sh#L102

Comment (debloper) [reply cico_run_tests.sh]: Please clarify the purpose, so that we're on the same page. Something seems off here.

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-56 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-57 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io

**Comment by jarifibrahim** [comment]
[test]

**Comment by jarifibrahim** [comment]
@pranavgore09 The `[test]` command triggered a new build. The `[test]` command is not project specific, it is `cico` specific.

**Comment by fabric8cd** [comment]
@pranavgore09 fabric8/fabric8-ui:SNAPSHOT-PR-2473-60 is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2473-fabric8-planner.badger.fabric8.io
